File: frontend/src/lib/spend.ts
/** Who this server has been working for.
 *
 *  `src/clickhouse/spend.rs` measures it. The diagnostics page already ranks
 *  statement *shapes* by cost, which answers what is expensive; this answers who
 *  it was expensive for, and that is usually the more actionable half. A shape
 *  costing forty minutes a week is a query to optimise. The same forty minutes
 *  belonging to one service account is a conversation with whoever owns it.
 *
 *  ## An empty account is not a person
 *
 *  ClickHouse logs work nobody asked for interactively — a materialized view's
 *  push, a subquery arriving from another node, a background flush — under an
 *  empty user. On the first server this was pointed at, that empty name was the
 *  *second largest spender on the machine*: 34% of the window, and 82% of its own
 *  time on one table. Named as a user it sends somebody hunting for an account
 *  that does not exist; named as what it is, it is one of the more useful things
 *  the reading produces, because a view quietly costing a third of a server is
 *  exactly the kind of thing nobody goes looking for.
 *
 *  ## And the window is the log's
 *
 *  Same rule as `lib/cold`, same vocabulary — `saysSpan` is imported rather than
 *  rewritten. A ranking of who spent the week, taken from a log holding five
 *  hours, is a ranking of who was awake this morning.
 */

import { saysSpan } from './cold'

export interface Spender {
  user: string
  /** True where the row is the server's own background work, not an account. */
  background: boolean
  statements: number
  seconds: number
  /** Of everything the window cost. */
  share: number
  read_bytes: number
  read_rows: number
  failed: number
  /** The table this account spent most of its time on. Empty where the log
   *  named none. */
  busiest_table: string
  /** That table's share of *this account's* time, not of the server's. */
  busiest_share: number
  last_seen: string
}

export interface SpendReport {
  available: boolean
  reason?: string
  window_days: number
  covered_days: number
  spenders: Spender[]
  total_seconds: number
  total_statements: number
  accounts: number
  /** False on a server that would not let Flint tag its own statements, where
   *  these figures include Flint reading them. */
  excludes_flint: boolean
}

/** Below this much log, a ranking of who spent the window is a ranking of who
 *  happened to be awake. The same floor `lib/cold` keeps, for the same reason. */
const NEEDS_DAYS = 1

/** And below this, there is no workload to divide up. */
const NEEDS_SECONDS = 10

export function trustworthy(report: SpendReport): { ok: boolean; why?: string } {
  if (!report.available) return { ok: false, why: report.reason ?? 'this reading is unavailable' }
  if (report.covered_days < NEEDS_DAYS) {
    return {
      ok: false,
      why: `system.query_log only goes back ${saysSpan(report.covered_days)}, which ranks who was awake this morning rather than who spends this server`,
    }
  }
  if (report.total_seconds < NEEDS_SECONDS) {
    return {
      ok: false,
      why: `the window holds ${Math.round(report.total_seconds)} seconds of query time altogether, which is not a workload to divide up`,
    }
  }
  return { ok: true }
}

/** How to name a row. Never "the user ''" — see the header. */
export function nameOf(spender: Spender): string {
  return spender.background ? 'The server’s own background work' : spender.user
}

/** The share above which one account is worth a sentence of its own.
 *
 *  A quarter, and it is a judgement: below it "who spends this server" has no
 *  answer worth reporting, because a server shared evenly between six accounts is
 *  a server working correctly. */
const NOTABLE = 0.25

/** The accounts worth saying something about, most expensive first.
 *
 *  Deliberately not "the top three": on a server where one account does
 *  everything, two of those three are noise, and on one shared evenly none of
 *  them is a finding. The threshold is the point.
 */
export function notable(report: SpendReport): Spender[] {
  if (!trustworthy(report).ok) return []
  return report.spenders.filter((s) => s.share >= NOTABLE)
}

/** The sentence for one account.
 *
 *  The busiest table is included only when it is most of what that account does.
 *  "41% of the server, and 12% of that on `events`" is two figures that together
 *  say nothing; "41% of the server, and 82% of that on one table" is a finding
 *  with somewhere to go.
 */
export function saysSpender(spender: Spender, report: SpendReport): string {
  const share = Math.round(spender.share * 100)
  const span = saysSpan(report.covered_days)
  const lead = `${nameOf(spender)} took ${share}% of this server’s query time over the last ${span}`
  const concentrated = spender.busiest_table !== '' && spender.busiest_share >= 0.5
  if (!concentrated) return `${lead}, across ${figure(spender.statements)} statements.`
  return `${lead}, and ${Math.round(spender.busiest_share * 100)}% of that was on one table — ${spender.busiest_table}.`
}

/** Statement counts, in the shape a sentence wants. `lib/format`'s `count` is
 *  for a column; this is prose, and "96.6 K statements" reads badly mid-clause. */
function figure(n: number): string {
  if (n < 1000) return String(n)
  if (n < 1_000_000) return `${Math.round(n / 1000)},000`
  return `${(n / 1_000_000).toFixed(1)} million`
}

/** The caveat that has to travel with these figures on a server Flint could not
 *  tag its own statements on. `null` where there is nothing to disclose. */
export function saysCaveat(report: SpendReport): string | null {
  if (!report.available || report.excludes_flint) return null
  return 'This server would not let Flint label its own statements, so Flint’s reading of the log is inside these figures.'
}

File: src/clickhouse/spend.rs
//! Who this server has been working for.
//!
//! `diagnostics::queries` already ranks statements by what they cost, which
//! answers *what* is expensive. It cannot answer *who*, and that is usually the
//! more actionable half: a statement shape costing forty minutes a week is a
//! query to optimise, and the same forty minutes belonging to one service
//! account is a conversation with whoever owns that service.
//!
//! One `GROUP BY user` over `system.query_log`, and three things it is careful
//! about.
//!
//! ## An empty user is not a person
//!
//! ClickHouse writes rows into `query_log` for work nobody asked for
//! interactively — a materialized view's push, a distributed subquery arriving
//! from another node, a background flush. Those carry an empty `user`, and on the
//! first server this was pointed at the empty name was the *largest spender*:
//! 414 seconds against 84 for the only real account on the machine. Reported as
//! a user it would send somebody looking for a person who does not exist, so it
//! is reported as what it is — the server's own background work — and it keeps
//! its figures, because that time is real and somebody should see it.
//!
//! ## A share needs a whole
//!
//! "41 minutes" means nothing without the total. Every row carries its share of
//! the window's own cost, computed on the server so the two cannot be summed
//! differently by two callers.
//!
//! ## Flint is left out, and says so
//!
//! By the tag every internal statement carries, exactly as the other query-log
//! readings do. On a server where Flint could not set `log_comment` — a
//! `readonly=1` profile — the tag is absent and Flint is *in* the figures; that
//! is what `settings_refused` on the session is for, and the page says it.

use serde::{Deserialize, Serialize};

use super::diagnostics::excluding_flint;
use super::{Client, QueryOptions, Reach};
use crate::error::Result;

/// One account, and what the server did for it.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Spender {
    /// The account as ClickHouse recorded it. Empty for the server's own
    /// background work — see `background`.
    pub user: String,
    /// True where `user` is empty: this row is a materialized view's push, a
    /// subquery from another node, a background flush. Not a person, and the
    /// caller must not name it as one.
    pub background: bool,
    pub statements: u64,
    /// Wall clock the server spent, in seconds.
    pub seconds: f64,
    /// That as a share of everything the window cost. Computed here so two
    /// callers cannot divide by two different totals.
    pub share: f64,
    pub read_bytes: u64,
    pub read_rows: u64,
    pub failed: u64,
    /// The table this account spent the most time on, and what that cost.
    ///
    /// Empty where the log names no table — a `SELECT 1`, a `SET`, a statement
    /// that failed before it was planned. The share is of *this account's* time,
    /// not of the server's: "most of what they do is on one table" is the
    /// sentence worth having.
    pub busiest_table: String,
    pub busiest_share: f64,
    /// When this account last ran anything.
    pub last_seen: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpendReport {
    pub available: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    pub window_days: u64,
    /// What the log can actually answer over, from its own oldest row. A ranking
    /// of who spent the week, over a log holding five hours, is a ranking of who
    /// was awake this morning.
    pub covered_days: f64,
    pub spenders: Vec<Spender>,
    /// Over every account, not only the listed ones.
    pub total_seconds: f64,
    pub total_statements: u64,
    pub accounts: u64,
    /// Whether Flint's own traffic could be excluded. False on a server that
    /// would not let Flint tag its statements, where these figures include Flint
    /// reading them.
    pub excludes_flint: bool,
}

/// Who the server worked for, most expensive first.
pub async fn spend(ch: &Client, days: u64, limit: u64) -> Result<SpendReport> {
    let days = days.clamp(1, 90);
    if let Some(why) = blocked(ch).await? {
        return Ok(SpendReport {
            available: false,
            reason: Some(why),
            window_days: days,
            covered_days: 0.0,
            spenders: Vec::new(),
            total_seconds: 0.0,
            total_statements: 0,
            accounts: 0,
            excludes_flint: false,
        });
    }

    let flint = excluding_flint(ch).await?;
    let excludes_flint = !flint.is_empty();

    /* One statement, with the window's own total riding inside it as a scalar
    subquery rather than fetched separately: two round trips over a busy log
    can straddle a flush, and a share computed against a total from another
    instant is a share that does not add up.

    The busiest table comes from `arrayJoin(tables)` per user, with each
    statement's whole duration attributed to every table it names. A join
    therefore contributes its full time to both sides, which overstates each
    — and it is the only apportionment the log supports: `query_log` records
    what a statement touched, never how its time divided between them. Said
    out loud on the field, because a reader comparing two figures deserves to
    know that one of them double-counts. */
    let sql = format!(
        "WITH whole AS ( \
             SELECT sum(query_duration_ms) AS ms, count() AS n, min(event_time) AS oldest, \
                    uniqExact(user) AS accounts \
             FROM system.query_log \
             WHERE event_time > now() - INTERVAL {days} DAY AND type = 'QueryFinish' {flint} \
         ), \
         /* A nested subquery rather than a second CTE, and not by taste: \
            ClickHouse substitutes a CTE by name, so `max(ms)` over a CTE whose \
            `ms` is itself `sum(...)` becomes `max(sum(...))` and is refused as \
            a nested aggregate. Inside a derived table the inner GROUP BY is \
            real and the outer one has an ordinary column to work on. */ \
         busiest AS ( \
             SELECT user, \
                    argMax(table, table_ms) AS table, \
                    max(table_ms)           AS ms \
             FROM ( \
                 SELECT user, \
                        arrayJoin(tables)       AS table, \
                        sum(query_duration_ms)  AS table_ms \
                 FROM system.query_log \
                 WHERE event_time > now() - INTERVAL {days} DAY \
                   AND type = 'QueryFinish' {flint} \
                   AND notEmpty(tables) \
                 GROUP BY user, table \
             ) \
             GROUP BY user \
         ) \
         SELECT q.user                                                 AS user, \
                /* `toBool`, because a comparison is a UInt8 and JSONEachRow \
                   writes it as 0/1 — which a `bool` field will not take. */ \
                toBool(q.user = '')                                    AS background, \
                toUInt64(count())                                      AS statements, \
                sum(q.query_duration_ms) / 1000                        AS seconds, \
                sum(q.query_duration_ms) / greatest(1, (SELECT ms FROM whole)) AS share, \
                toUInt64(sum(q.read_bytes))                            AS read_bytes, \
                toUInt64(sum(q.read_rows))                             AS read_rows, \
                toUInt64(countIf(q.exception_code != 0))               AS failed, \
                any(coalesce(b.table, ''))                             AS busiest_table, \
                any(coalesce(b.ms, 0)) / greatest(1, sum(q.query_duration_ms)) AS busiest_share, \
                toString(max(q.event_time))                            AS last_seen \
         FROM system.query_log AS q \
         LEFT JOIN busiest AS b ON b.user = q.user \
         WHERE q.event_time > now() - INTERVAL {days} DAY \
           AND q.type = 'QueryFinish' {flint} \
         GROUP BY q.user \
         ORDER BY seconds DESC \
         LIMIT {}",
        limit.clamp(1, 100)
    );

    let opts = QueryOptions {
        quote_64bit_integers: false,
        introspection: true,
        ..Default::default()
    };
    let spenders: Vec<Spender> = ch.rows_with(&sql, opts.clone()).await?;

    #[derive(Deserialize)]
    struct Whole {
        ms: f64,
        n: u64,
        covered_days: f64,
        accounts: u64,
    }
    let whole: Option<Whole> = ch
        .row_with(
            &format!(
                "SELECT sum(query_duration_ms)                          AS ms, \
                        toUInt64(count())                               AS n, \
                        toUInt64(uniqExact(user))                       AS accounts, \
                        least({days}.0, dateDiff('second', min(event_time), now()) / 86400.0) \
                                                                        AS covered_days \
                 FROM system.query_log \
                 WHERE event_time > now() - INTERVAL {days} DAY \
                   AND type = 'QueryFinish' {flint}"
            ),
            opts,
        )
        .await?;

    Ok(SpendReport {
        available: true,
        reason: None,
        window_days: days,
        covered_days: whole
            .as_ref()
            .map(|w| w.covered_days.max(0.0))
            .unwrap_or(0.0),
        total_seconds: whole.as_ref().map(|w| w.ms / 1000.0).unwrap_or(0.0),
        total_statements: whole.as_ref().map(|w| w.n).unwrap_or(0),
        accounts: whole.map(|w| w.accounts).unwrap_or(0),
        spenders,
        excludes_flint,
    })
}

async fn blocked(ch: &Client) -> Result<Option<String>> {
    Ok(match ch.reach("query_log").await? {
        Reach::Readable => None,
        Reach::Denied => Some("this user is not granted SELECT on system.query_log".to_string()),
        Reach::Absent | Reach::Unconfigured => {
            Some("system.query_log is not enabled on this server".to_string())
        }
    })
}
